src/model_lifecycle.cc

```diff
@@ -0,0 +1,761 @@
+  }
+
+  std::set<int64_t> versions;
+  RETURN_IF_ERROR(
+      VersionsToLoad(model_path, model_name, model_config, &versions));
+  if (versions.empty()) {
+    return Status(
+        Status::Code::INVALID_ARG,
+        "at least one version must be available under the version policy of "
+        "model '" +
+            model_name + "'");
+  }
+
+
+  const uint64_t now_ns =
+      std::chrono::duration_cast<std::chrono::nanoseconds>(
+          std::chrono::steady_clock::now().time_since_epoch())
+          .count();
+  std::shared_ptr<LoadTracker> load_tracker(
+      new LoadTracker(versions.size(), now_ns));
+  for (const auto& version : versions) {
+    std::unique_ptr<ModelInfo> linfo(
+        new ModelInfo(model_path, model_config, now_ns));
+    ModelInfo* model_info = linfo.get();
+
+    LOG_INFO << "loading: " << model_name << ":" << version;
+    model_info->state_ = ModelReadyState::LOADING;
+    model_info->state_reason_.clear();
+    model_info->agent_model_list_ = agent_model_list;
+
+    auto res = it->second.emplace(
+        std::make_pair(version, std::unique_ptr<ModelInfo>()));
+    if (res.second) {
+      res.first->second = std::move(linfo);
+    } else {
+      // There is already a record of this model version. Check if the version
+      // model is being served, if so, the re-load of the version
+      // should be performed in background to avoid version downtime.
+      // Otherwise, swap and monitor state for newly loading model.
+      auto& serving_model = res.first->second;
+      std::lock_guard<std::recursive_mutex> lock(serving_model->mtx_);
+      if (serving_model->state_ == ModelReadyState::READY) {
+        background_models_[(uintptr_t)model_info] = std::move(linfo);
+      } else {
+        // swap the monitoring model info
+        serving_model.swap(linfo);
+
+        // further check the state, put to 'background_models_' to keep
+        // the object valid if the model is LOADING / UNLOADING, because
+        // the model info will be accessed by a different thread once the
+        // operation is completed
+        if ((linfo->state_ == ModelReadyState::LOADING) ||
+            (linfo->state_ == ModelReadyState::UNLOADING)) {
+          ModelInfo* key = linfo.get();
+          background_models_[(uintptr_t)key] = std::move(linfo);
+        }
+      }
+    }
+
+    // Load model asynchronously via thread pool
+    load_pool_->Enqueue([this, model_name, version, model_info, OnComplete,
+                         load_tracker]() {
+      CreateModel(model_name, version, model_info);
+      OnLoadComplete(model_name, version, model_info, OnComplete, load_tracker);
+    });
+  }
+
+  return Status::Success;
+}
+
+void
+ModelLifeCycle::CreateModel(
+    const std::string& model_name, const int64_t version, ModelInfo* model_info)
+{
+  LOG_VERBOSE(2) << "CreateModel() '" << model_name << "' version " << version;
+  const auto& model_config = model_info->model_config_;
```

Comment:
There used to be a [lock](https://github.com/triton-inference-server/core/blob/6a984dd747df4b0cc1782f63169a16ef9eb1ea7c/src/model_repository_manager.cc#L1259-L1263) on the model_info->mtx here before getting the model_config - is this no longer necessary?

Reply:
I didn't state this explicitly, in the current implementation, [some members](https://github.com/triton-inference-server/core/pull/92/files#diff-5df42a57c62d9f5cf01af9061da771a5e827f41e746e8c777ce021de3a14e5d7R207-R209) in `ModelInfo` will remain unchanged, so we don't really need a lock around those members. In previous implementation, this is not the case because we may reuse the `ModelInfo`